### PR TITLE
New Feature: Path based parameters

### PR DESCRIPTION
**Pull request by Tiberriver256**
*Note this should also fix the matching issue described at the bottom of issue #116 

This allows for the following scenario:
```ps
New-PolarisGetRoute -Path "/Hello/{Name}" -ScriptBlock {
                $Response.Send("Hello $($Parameters.Name)")
            }

$Result = Invoke-WebRequest -Uri "http://localhost:$Port/hello/Testing" -UseBasicParsing
$Result.Content | Should Be 'Hello Testing'
```

It introduces a new $Parameters PSCustomObject representing any path based parameters and it also makes route matching more strict as it is now converts paths to a regular expression using Convert-PathParameters function.

At the time a request is made the above mentioned path ```/Hello/{Name}``` would get converted to a regular expression of ```^/Hello/(?<Name>.*)$``` for matching against the client's requested route.

A more basic example would be that ```/Hello``` would be converted to ```^/Hello$``` to guarantee an exact path match. I've also added some test cases to ensure we are throwing 404 errors for paths that don't match exactly.

### CODE REVIEW

**Comment by Tiberriver256** [comment]
Appveyor is still drunk, but tests appear to be passing. Take a look at the changes and let me know what you think @tylerl0706 

https://ci.appveyor.com/project/PowerShell/polaris/build/job/0qxc0cy0p5bd54g9

**Comment by Tiberriver256** [comment]
Thinking about this a bit more last night. If you like the concept I may just look into porting the module expressjs uses. It would give a more familiar syntax and theirs is pretty robust.

https://github.com/pillarjs/path-to-regexp

**Comment by tylerl0706** [comment]
@Tiberriver256 THIS IS AWESOME AND EXACTLY WHAT POLARIS NEEDS. Adopting what expressjs does sounds awesome for this - that's one of my favorite features of express actually.

**Comment by tylerl0706** [comment]
Do you still want me to review this or do you want to take a crack at following what expressjs does 😀

**Comment by Tiberriver256** [comment]
I’ll take a crack at the expressjs :) doesn’t look too bad.

**Comment by TimCurwick** [comment]
Nice enhancement, but the implementation makes certain assumptions about our desired end state that may make sense in the context of the this enhancement, but may not be valid for other desired behaviors. I would prefer if we take the time to discuss and agree on a set of behaviors and design requirements, so that we can design an elegant, comprehensive solution, rather than hoping we get there through a series of small breaking bug fixes and enhancements taking us in random directions.

**Comment by Tiberriver256** [comment]
@TimCurwick - Agreed, I think the goal of my PR here (which was to get the discussion started) is not really standard. I opened #120 to start the discussion. Really glad people are getting into this!

**Comment by Tiberriver256** [comment]
I updated the code here as a first stab at RFC001 (#120). Everything is working and has a test case for the examples I gave. It does not have a method for switching from case sensitive to being insensitive yet.

*EDIT: Looks like I've actually got one test to fix on the linux side but the code should still be okay to look at.

File: lib/Polaris.Class.ps1 [hunk 1]
```diff
@@ -48,7 +48,8 @@ class Polaris {
                 # Run middleware in the order in which it was added
                 foreach ($Middleware in $Polaris.RouteMiddleware) {
 					$InformationVariable += $Polaris.InvokeRoute(
-							$Middleware.Scriptblock,
+                            $Middleware.Scriptblock,
+                            $Null,
```
Comment (tylerl0706) [inline lib/Polaris.Class.ps1]: nit: some weird indenting!

File: lib/Polaris.Class.ps1 [hunk 2]
```diff
@@ -68,7 +75,8 @@ class Polaris {
                     try {
 
                         $InformationVariable += $Polaris.InvokeRoute(
-							$Routes[$MatchingRoute][$Request.Method],
+                            $Routes[$MatchingRoute][$Request.Method],
+                            $Parameters,
```
Comment (tylerl0706) [inline lib/Polaris.Class.ps1]: nit some more weird indents

File: lib/Polaris.Class.ps1 [hunk 3]
```diff
@@ -192,6 +201,23 @@ class Polaris {
         return $SanitizedPath
     }
 
+    static [RegEx] ConvertPathToRegex([string]$Path){
+        Write-Debug "Path: $path"
+        $path = $path -replace '\.', '\.'
```
Comment (tylerl0706) [inline lib/Polaris.Class.ps1]: Is this replace doing anything?

File: lib/Polaris.Class.ps1 [hunk 4]
```diff
@@ -192,6 +201,23 @@ class Polaris {
...
+        $path = $path -replace '-', '\-'
+        $path = $path -replace '\*', '.*'
+        $path = "^$path$"
+        $path = $path -replace ":(\w+)(\?{0,1})", '(?<$1>.+)$2'
```
Comment (tylerl0706) [inline lib/Polaris.Class.ps1]: can you add a couple comments in this function due to the complexity?

File: lib/Polaris.Class.ps1 [hunk 5]
```diff
@@ -57,7 +58,13 @@ class Polaris {
                 $Polaris.Log("Parsed Route: $Route")
                 $Polaris.Log("Request Method: $($RawRequest.HttpMethod)")
                 $Routes = $Polaris.ScriptblockRoutes
-                $MatchingRoute = $Routes.keys.Where( { $Route -match $_ })[0]
+
+                #
+                # Searching for the first route that matches by the most specific route paths first.
+                #
+                $MatchingRoute = $Routes.keys | Sort-Object -Property Length -Descending | Where-Object { $Route -match [Polaris]::ConvertPathToRegex($_) } | Select-Object -First 1
+                $Request.Parameters = ([PSCustomObject]$Matches)
```
Comment (tylerl0706) [inline lib/Polaris.Class.ps1]: Where does `$Matches` get defined?

**Comment by tylerl0706** [comment]
Also that test failure is bogus...

```
Expected strings to be the same, but they were different.
String lengths are both 47.
Strings differ at index 2.
Expected: '@{bookId=123; userId=12; 0=/users/12/books/123}'
But was:  '@{userId=12; bookId=123; 0=/users/12/books/123}'
```

Maybe do a different kind of comparison rather than a string compare.

File: lib/Polaris.Class.ps1 (hunk 5 above)
Comment (Tiberriver256) [reply lib/Polaris.Class.ps1]: $Matches is an automatic variable that gets defined every time you run a -match comparison. 

See:
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-6#matches
https://mcpmag.com/articles/2015/09/30/regex-groups-with-powershell.aspx

**Comment by Tiberriver256** [comment]
Yeah, no worries on the timing. My life kind of blew up for a bit there too but seems to be close to returning to normalcy. It might be a bit before I can get to cleaning this up but super excited to do it.

**Comment by tylerl0706** [comment]
@Tiberriver256 Got all the release stuff done so as soon as we get this PR in I'll throw Polaris up on the Gallery!

**Comment by tylerl0706** [comment]
Gentle ping 😊

I'd like to get this on the Gallery.

**Comment by Tiberriver256** [comment]
Thanks Tyler. I haven't forgotten. I'll see if I can get this up tonight.

**Comment by Tiberriver256** [comment]
Should be all set sir :)

**Comment by tylerl0706** [comment]
Whooooooo!!

**Pull request merged**